crates/bevy_ecs/src/event/base.rs

```diff
@@ -421,3 +424,28 @@ pub(crate) struct EventInstance<E: BufferedEvent> {
     pub event_id: EventId<E>,
     pub event: E,
 }
+
+/// # Warning
+///
+/// This struct should only be instantiated internally to this crate.
+///
+/// [`EventKey`]s are created in:
+/// - `crate::observer::trigger_dynamic_ref_with_caller`
+/// - `crate::observer::trigger_dynamic_ref_with_caller`
+/// - `crate::observer::observer_multiple_events`
+/// - `crate::observer::observer_dynamic_trigger`
+/// - `crate::observer::runner::hook_on_add`
```

Comment:
```suggestion
/// A unique identifier for an [`Event`], used by [observers].
///
/// You can look up the key for your event by calling the [`Event::event_key`] method.
///
/// [observers]: crate::observer
```

We don't need to give a warning about non-public construction: Rust makes that impossible with correct usage of the visibility system. But we should give some docs about what this type is for and how to get it.